### PR TITLE
WIP: DNM: arm: Add ramfunc section to link scripts

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
There may has some code that can only run in sram such as the
operations about flash, so introduce a method that placing
code into sram, load from flash to sram when booting.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

### CODE REVIEW

**Comment by qianfan-Zhao** [comment]
Actually this link scripts doesn't  work. I write a simple code to test:

```
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index 2d4a7f0..f4ad993 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -6,8 +6,18 @@
 
 #include <zephyr.h>
 #include <misc/printk.h>
+#include <linker/linker-defs.h>
+
+static void __attribute__((noinline)) _GENERIC_SECTION(ramfunc) ramfunc_test(void)
+{
+       printk("__ramfunc_ram_start: %p\n", __ramfunc_ram_start);
+       printk("__ramfunc_ram_end: %p\n", __ramfunc_ram_end);
+       printk("__ramfunc_rom_start: %p\n", __ramfunc_rom_start);
+       printk("The address of %s is %p\n", __func__, ramfunc_test);
+}
 
 void main(void)
 {
        printk("Hello World! %s\n", CONFIG_BOARD);
+       ramfunc_test();
 }
```

Compile and download this demo to nucleo_f070rb, next is the address printed from console:

```
***** Booting Zephyr OS v1.13.0-rc1-623-g19ff1d2 *****
Hello World! nucleo_f070rb
__ramfunc_ram_start: 0x2000113c
__ramfunc_ram_end: 0x2000113c
__ramfunc_rom_start: 0x08002240
The address of ramfunc_test is 0x08001d7d
```

The function 'ramfunc_test' linked into 'ramfunc section' but the address is in FLASH region.

```
➜  build git:(ramfunc) ✗ arm-none-eabi-objdump -jramfunc -S  zephyr/zephyr.elf

zephyr/zephyr.elf:     file format elf32-littlearm


Disassembly of section ramfunc:

08001d7c <ramfunc_test>:
#include <zephyr.h>
#include <misc/printk.h>
#include <linker/linker-defs.h>

static void __attribute__((noinline)) _GENERIC_SECTION(ramfunc) ramfunc_test(void)
{
 8001d7c:       b510            push    {r4, lr}
        printk("__ramfunc_ram_start: %p\n", __ramfunc_ram_start);
 8001d7e:       4909            ldr     r1, [pc, #36]   ; (8001da4 <_image_text_end+0x28>)
 8001d80:       4809            ldr     r0, [pc, #36]   ; (8001da8 <_image_text_end+0x2c>)
 8001d82:       f7fe fc8f       bl      80006a4 <printk>
        printk("__ramfunc_ram_end: %p\n", __ramfunc_ram_end);
 8001d86:       4909            ldr     r1, [pc, #36]   ; (8001dac <_image_text_end+0x30>)
 8001d88:       4809            ldr     r0, [pc, #36]   ; (8001db0 <_image_text_end+0x34>)
 8001d8a:       f7fe fc8b       bl      80006a4 <printk>
        printk("__ramfunc_rom_start: %p\n", __ramfunc_rom_start);
 8001d8e:       4909            ldr     r1, [pc, #36]   ; (8001db4 <_image_text_end+0x38>)
 8001d90:       4809            ldr     r0, [pc, #36]   ; (8001db8 <_image_text_end+0x3c>)
 8001d92:       f7fe fc87       bl      80006a4 <printk>
        printk("The address of %s is %p\n", __func__, ramfunc_test);
 8001d96:       4a09            ldr     r2, [pc, #36]   ; (8001dbc <_image_text_end+0x40>)
 8001d98:       4909            ldr     r1, [pc, #36]   ; (8001dc0 <_image_text_end+0x44>)
 8001d9a:       480a            ldr     r0, [pc, #40]   ; (8001dc4 <_image_text_end+0x48>)
 8001d9c:       f7fe fc82       bl      80006a4 <printk>
}
 8001da0:       bd10            pop     {r4, pc}
 8001da2:       46c0            nop                     ; (mov r8, r8)
 8001da4:       2000113c        .word   0x2000113c
 8001da8:       08001e84        .word   0x08001e84
 8001dac:       2000113c        .word   0x2000113c
 8001db0:       08001e9d        .word   0x08001e9d
 8001db4:       08002240        .word   0x08002240
 8001db8:       08001eb4        .word   0x08001eb4
 8001dbc:       08001d7d        .word   0x08001d7d
 8001dc0:       08001e58        .word   0x08001e58
 8001dc4:       08001ecd        .word   0x08001ecd
```

**Comment by qianfan-Zhao** [comment]
I had searched 'how to place code into sram' on google, and find a way: [ramfunc-gcc](https://mcu-things.com/blog/ramfunc-gcc/)

```
longcall：This is an attribute for function. It declares that the function with this attribute is to be called from different place. The compiler always uses a pointer to call the function.

section ：This attribute places variables and/or functions in section you specify.

With these attributes specified, you can place your functions or variables in RAM. If you want a variable to be placed in RAM, you may not use longcall attribute. That attribute requires for function, but not for variables.

Code example :  __attribute__ ((longcall, section(“memory_section”)))
“memory_section” is the section in which you want your function to be placed.
```

And changed the **ramfunc_test** in this way:

```
static void __attribute__((noinline)) __attribute__((long_call,section(".ramfunc"))) ramfunc_test(void)
```

Now everything is OK.

```
***** Booting Zephyr OS v1.13.0-rc1-623-g19ff1d2 *****
Hello World! nucleo_f070rb
__ramfunc_ram_start: 0x20001140
__ramfunc_ram_end: 0x200011a0
__ramfunc_rom_start: 0x080021fc
The address of ramfunc_test is 0x20001141
```

**Comment by qianfan-Zhao** [comment]
@SebastianBoe could you please also check it?

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
Hi,

I am not able to get this to work on nrf52840_pca10056.

I don't have a nucleo to test with.

```
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index 2d4a7f0d5f..c86387dd9f 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -7,7 +7,13 @@
 #include <zephyr.h>
 #include <misc/printk.h>
 
+void __ramfunc function_running_from_ram(void)
+{
+       printk("Running %d\n", (int)&function_running_from_ram);
+}
+
 void main(void)
 {
        printk("Hello World! %s\n", CONFIG_BOARD);
+       function_running_from_ram();
 }

```

```
***** Booting Zephyr OS zephyr-v1.13.0-389-ga90f435ffb *****
Hello World! nrf52840_pca10056
***** MPU FAULT *****
  Instruction Access Violation
***** Hardware exception *****
Current thread ID = 0x20000d48
Faulting instruction address = 0x20002aa8
Fatal fault in essential thread! Spinning...

```

```
20002aa4 <function_running_from_ram>:
	printk("Running %d\n", (int)&function_running_from_ram);
20002aa4:	4901      	ldr	r1, [pc, #4]	; (20002aac <function_running_from_ram+0x8>)
20002aa6:	4802      	ldr	r0, [pc, #8]	; (20002ab0 <function_running_from_ram+0xc>)
20002aa8:	f000 b806 	b.w	20002ab8 <__printk_veneer>
20002aac:	20002aa5 	.word	0x20002aa5
20002ab0:	0000ae8b 	.word	0x0000ae8b
20002ab4:	00000000 	.word	0x00000000

20002ab8 <__printk_veneer>:
20002ab8:	f85f f000 	ldr.w	pc, [pc]	; 20002abc <__printk_veneer+0x4>
20002abc:	00000dcd 	.word	0x00000dcd

```

**Comment by SebastianBoe** [comment]
Another thing.

What this approach does is introduce an output section, an input section, and kernel code for init'ing the new output section.

We can make do with a lot less code if we introduce a new input section and re-use an existing output section, like for instance the DATA output section.

```
diff --git a/include/arch/arm/cortex_m/scripts/linker.ld b/include/arch/arm/cortex_m/scripts/linker.ld
index c9baa7d9fc..afcd85cf5b 100644
--- a/include/arch/arm/cortex_m/scripts/linker.ld
+++ b/include/arch/arm/cortex_m/scripts/linker.ld
@@ -336,6 +336,7 @@ SECTIONS
 	__data_ram_start = .;
 	KERNEL_INPUT_SECTION(.data)
 	KERNEL_INPUT_SECTION(".data.*")
+	KERNEL_INPUT_SECTION(".ramfunc")
 	*(".kernel.*")
 
 #ifdef CONFIG_CUSTOM_RWDATA_LD

```

**Comment by qianfan-Zhao** [comment]
@SebastianBoe I had test this pr on different platform one has MPU and the other doesn't. this pr can't run on MPU platform, I think maybe can't run code on ram due to mpu protected. I will check and fix it.


Placed ramfunc section to data is a good solution that can make things easier, will not copy ramfunc from flash to ram, the original code do that. the biggest problem is how to deal with mpu, I don't have any idea now .

**Comment by qianfan-Zhao** [comment]
@SebastianBoe 
ARM soc has 3 MPU region default in Zephyr, the whole SRAM(including SRAM_0 and SRAM_1) are configured as XN(eXecuted Nerver) default. So make code run in SRAM will cause MPU fault.

```
static struct arm_mpu_region mpu_regions[] = {
	/* Region 0 */
	MPU_REGION_ENTRY("FLASH_0",
			 CONFIG_FLASH_BASE_ADDRESS,
			 REGION_FLASH_ATTR(REGION_FLASH_SIZE)),

	MPU_REGION_ENTRY("SRAM_0",
			 CONFIG_SRAM_BASE_ADDRESS,
			 REGION_RAM_ATTR(REGION_SRAM_0_SIZE)),

#ifdef REGION_SRAM_1_SIZE
	MPU_REGION_ENTRY("SRAM_1",
			 CONFIG_SRAM_BASE_ADDRESS + REGION_SRAM_1_START,
			 REGION_RAM_ATTR(REGION_SRAM_1_SIZE)),
#endif
};

struct arm_mpu_config mpu_config = {
	.num_regions = ARRAY_SIZE(mpu_regions),
	.mpu_regions = mpu_regions,
};
```

```
/* Some helper defines for common regions */
#define REGION_RAM_ATTR(size) \
{ \
	(NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NON_SHAREABLE | \
	 MPU_RASR_XN_Msk | size | P_RW_U_NA_Msk) \
}
```

If I remove **MPU_RASR_XN_Msk** in REGION_RAM_ATTR, the **ramfunc** are runable.

But I don't think mark the whole SRAM region as executed is a good idea, is there has another better way to do that?

**Comment by SebastianBoe** [comment]
I guess we could partition the ram into an executable region and a non-executable region and have the MPU just protect the non-executable region.

**Comment by qianfan-Zhao** [comment]
@SebastianBoe 
agree. I will read the mpu part and find a better way to implement.
That's will take many days i think.

File: include/toolchain/gcc.h [hunk 1]
```diff
@@ -113,6 +113,11 @@ do {                                                                    \
 
 #define __in_section_unique(seg) ___in_section(seg, __FILE__, __COUNTER__)
 
+/* Using this places a function in RAM instead of FLASH.
+ */
+#define __ramfunc	__attribute__((noinline))			\
```
Comment (pfalcon) [inline include/toolchain/gcc.h]: We cannot use identifiers starting with double undescores. cc: @ceolin

I'd recommend to follow standard preprocessor macro convention and call it something like RAMFUNC, but seems we already have things like `_app_dmem(id)` and `_app_bmem(id)` ;-/ , so `_ramfunc` would do...

Comment (qianfan-Zhao) [reply include/toolchain/gcc.h]: Or maybe something like _GENERIC_SECTION()?

Comment (qianfan-Zhao) [reply include/toolchain/gcc.h]: There has some __attribute__ marco starting with double underscores, such as:
__weak, __used, __unused, __packed, __aligned, So I prefer using __ramfunc.

**Comment by easonxiang** [comment]
@qianfan-Zhao I'm facing the same problem, do you have a better way to solve this issue?

**Comment by qianfan-Zhao** [comment]
@easonxiang I have no idea about MPU current, so I just closed this PR.

Seems the only way to solve this is create a standalone link section in SRAM and set it as executed in MPU.

**Comment by qianfan-Zhao** [comment]
Closed this pr and open a new one: #11228

**Pull request closed without merging**